### ISSUE TITLE
DSCP option isn't actually marking packets

### ISSUE THREAD

**Issue by squircle** [rclone/rclone#5293]
The DSCP marks functionality added in v1.55.0 doesn't seem to be working at all. @xenago and I tested on the following platform/version/backend combinations, and rclone didn't produce DSCP-marked packets on any of them:
* macOS 10.14 (Mojave), rclone 1.55.1, Google Drive IPv4 & IPv6
* Windows 10 20H2, rclone 1.55.1, Google Drive IPv4, Onedrive (personal) IPv4
* OpenSUSE Tumbleweed, rclone 1.55.0, Google Drive IPv4 & IPv6

I originally discovered the problem while using the restic integration, but the behaviour is the same on a simple copy too. I also used iperf3 to verify that DSCP tags weren't being mangled/washed by the network stack, and iperf3 emitted DSCP-tagged packets as expected.

**How to reproduce:**

1. Start tcpdump/Wireshark/tshark
2. Invoke rclone with the `--dscp LE` argument (works for all possible values, not just `LE`). For example: `rclone --dscp LE copy testfile remote:`. Optionally, add `--bind <address>` to force IPv4 on a dual-stack host.
3. After rclone finishes, check the DiffServ field of an outgoing packet to see that it isn't being marked (CS0). In Wireshark, this shows as `Differentiated Services Field: 0x00 (DSCP: CS0, ECN: Not-ECT)` for IPv4, and `Traffic Class: 0x00 (DSCP: CS0, ECN: Not-ECT)` for IPv6.

I've attached the following:
* A packet capture from the macOS example (IPv6) [rclone-dscp-bug-mac-v6.pcapng.gz](https://github.com/rclone/rclone/files/6402470/rclone-dscp-bug-mac-v6.pcapng.gz)
* The corresponding debug log [rclone-dscp-bug-mac.log](https://github.com/rclone/rclone/files/6402471/rclone-dscp-bug-mac.log)

Neither of us is familiar with golang, but I'll try to get a debug build running to narrow down the problem. I wanted to put it out there in case somebody more familiar with rclone/golang can spot the error faster.

**Edit:**

I put it into a debugger, and the problem appears to be in `dialer.go`: https://github.com/rclone/rclone/blob/9e2fbe0f1a9314e1330b4adc9d69de137be12ee6/fs/fshttp/dialer.go#L48-L68

I set breakpoints at lines 55-57. 55 and 56 are called as expected, but **line 57 is never reached**. Therefore, `SetTrafficClass`/`SetTOS` are never called.

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by squircle** [comment]
Adding the print statement gives `addr = &net.TCPAddr{IP:net.IP{0x26, 0x7, 0xf8, 0xb0, 0x40, 0xb, 0x8, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x20, 0xa}, Port:443, Zone:""}`. I'm using the Google Drive backend for my testing, but the behaviour is the same with the Onedrive backend.

**Comment by squircle** [comment]
Ah-ha! If I reduce the if statement to just an assignment (`addr := c.RemoteAddr().(*net.IPAddr)`), I get this:

```
panic: interface conversion: net.Addr is *net.TCPAddr, not *net.IPAddr
```

So it seems like a type mismatch. Changing `c.RemoteAddr().(*net.IPAddr)` to `c.RemoteAddr().(*net.TCPAddr)` fixes the problem and emits DSCP-marked packets on IPv6 and IPv4 (tested on macOS). I can send a PR for this if it's the "right" solution after testing on Windows and Linux as well.

**Comment by ncw** [comment]
I was hoping it would be something like that!

Changing it to *TCPAddr will fix it for all TCP connections which is great!

It is possible that there will be other types of connection coming through here but at the moment they will all be TCP so that should work fine.

Something like this (untested) would be future proof

```go
		var ip net.Addr
		switch addr := c.RemoteAddr().(type) {
		case *net.IPAddr:
			ip = addr.IP
		case *net.TCPAddr:
			ip = addr.IP
		case *net.UDPAddr:
			ip = addr.IP
		default:
		}
		if addr != nil {
			if addr.IP.To16() != nil && addr.IP.To4() == nil {
				err = ipv6.NewConn(c).SetTrafficClass(d.tclass)
			} else {
				err = ipv4.NewConn(c).SetTOS(d.tclass)
			}
			if err != nil {
				return c, err
			}
		}
```

However it is only doing that in order to figure out whether the connection is IPv4 or IPv6 and I think you could do that like this without all that complication

```go
			isIPv6 := strings.Count(c.RemoteAddr().String(), ":") > 1 // IPv6 addresses must have 2 or more ':'
			if isIPv6 {
				err = ipv6.NewConn(c).SetTrafficClass(d.tclass)
			} else {
				err = ipv4.NewConn(c).SetTOS(d.tclass)
			}
			if err != nil {
				return c, err
			}

```

**Comment by squircle** [comment]
Thanks for those blurbs. I've made that change in https://github.com/squircle/rclone/commit/ef0ea13022f3378fb12efa785dc1143aab998240 and done the following testing:
* macOS IPv4: working ✅
* macOS IPv6: working ✅
* openSUSE Tumbleweed IPv4: working ✅
* openSUSE Tumbleweed IPv6: working ✅
* Windows 10 20H2 IPv4: fails silently ❌
* Windows 10 20H2 IPv6: crashes with "not implemented" error ❌

It's lucky that DSCP wasn't working at all on previous releases, because it's totally broken on Windows. Based on https://github.com/golang/go/issues/42728 it seems that Microsoft only supports setting DSCP through a proprietary Windows QoS API and not through socket options. For that reason, I'm going to update `dialer.go` (and the related documentation) to skip DSCP handling on Windows platforms. I'll make it easy to refactor if the golang packages add support for this API (or if somebody else wants to implement this).

I'll send a PR with the whole set of changes once I'm done.

**Comment by ncw** [comment]
Great testing - well done.

We could call those APIs anyway under Windows and detect the errors and ignore in the hope that the Go team will fix them up one day?

Or maybe you have a better suggestion?

We could also log something (but only once so we'd want a sync.Once gate in there).

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.56

Thank you @squircle for fixing the problem.
